tests/types/test_modules.py

```diff
@@ -0,0 +1,155 @@
+# The following tests ensure that the resolver is able to resolve all types
+# using the correct module. Concrete types should be non-problematic and are
+# only included here for completeless. The problematic cases are where a
+# `str` is given as type which can only be resolved at schema construction.
+#
+# Important: For this tests not to hide any wrong resolves no types must
+# be imported into the namespaces under the same name and all type names
+# must be unique. e.g. when importing `a.AObject` into the `c` module it
+# must be renamed to something like `C_AObject`. That way wrong resolves
+# like `C_AObject` being loaded from `a` or `AObject` being loaded from
+# `c` can be detected.
+
+
+from typing import List
+
+from modules import a_mod, b_mod, c_mod, x_mod
```

Comment:
Would it be possible to rename these a/b modules to be a bit more self-describing? It's not clear just looking a the test names what the difference between a/b/c/x are and what the exact purpose of each test is.

Reply:
Also, what do you think about moving the test file to be within the `modules` directory, so that the "importing resolvers from other modules" tests are right with their respective files?

Reply:
A and B are standalone modules. C has the types doing composition and inheritance (using types from A and B). And X ... well... it's the unknown containing the resolver without typing. My first instinct was to go with `foo`, `bar` and `baz` but I soon realized that `a`, `b` and `c` is actually easier to reason about especially as all types are prefixed with that letter. Going with something more descriptive would just increase the already long type names. Do you have a better idea for the module names and type names? Feel free to change it. I don't have a better idea. :see_no_evil: 

I'm fine with moving the `test_modules.py` into the `modules` directory. I just didn't do it as it felt like it was just a supplement to the `test_modules.py` file.

Reply:
Hmm, I see what you're saying. Not really good options for naming. Can you add `docstring` at the top of each file explaining what's inside, as an alternative.

> I'm fine with moving the test_modules.py into the modules directory. I just didn't do it as it felt like it was just a supplement to the test_modules.py file.

I think making the test entirely self-contained in a directory would be best.

Reply:
I just added a docstring to the `__init__.py` explaining the files. I can move those strings to every module if that's better.

Reply:
How about a `README.md` file explaining the whole purpose of the module in the tests?

That could actually also include the part of the `test_modules.py` docstring explaining the reason why the types are named that way.

Reply:
> I just added a docstring to the __init__.py explaining the files. I can move those strings to every module if that's better.

I think that would be best. I don't know if we even need to have the `__init__.py`. The tests work fine without it

Reply:
That `c.C_AObject` alias for the `a.AObject` surely isn't self explaining and someone might want to refactor this in the future as it's just a "silly rename".

Reply:
> How about a README.md file explaining the whole purpose of the module in the tests?

I have no preference here, so I'll leave the discretion to you.

Reply:
Did we decide to not move the tests here to a self-enclosed directory?

Reply:
I can move the `test_modules.py` into the `modules` directory, but it will require two changes:

1. The directory `modules` needs a `__init__.py` file
2. The imports from `test_modules.py` need to be changed to `from . import a_mod, b_mod, c_mod, x_mod`

OR

All `*._mod.py` files are moved into their own directory and the content is moved into their `__init__.py`.

It seams that absolute module imports at root level only work for directories when running tests.

Reply:
This happens when I try to run it using an absolute import.
```
$ poetry run pytest tests/types/modules/test_modules.py 
(...)
_________________________ ERROR collecting tests/types/modules/test_modules.py _____________________________________________________________________________
ImportError while importing test module '/home/bikeshedder/projects/foss/strawberry/tests/types/modules/test_modules.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../../.asdf/installs/python/3.8.6/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/types/modules/test_modules.py:11: in <module>
    import a_mod, b_mod, c_mod, x_mod
E   ModuleNotFoundError: No module named 'a_mod'
```
Only when switching to a relative import and adding a `__init__.py` I'm able to run the tests.

Reply:
I think it is fine to use relative and `__init__.py` 😄

Reply:
Now the file lives in the `tests/types/models` directory. :+1: